google/cloud/datastore/batch.py

```diff
@@ -218,7 +231,7 @@ def put(self, entity):
                  progress, if entity has no key assigned, or if the key's
                  ``project`` does not match ours.
         """
-        if self._status != self._IN_PROGRESS:
+        if not self._allow_mutations():
```

Comment:
The [current implementation](https://github.com/googleapis/python-datastore/blob/80612021b82d1b62d0271107f73571cde8262256/google/cloud/datastore/transaction.py#L374) behaves as you describe when the `begin_later` flag is True. When it's False, it currently falls back to the old expectations, where any action before calling begin is treated as an error (because begin was expected to be called as part of the init. If it wasn't, something is wrong)

It sounds like you're suggesting we should remove these extra validation checks when `begin_later=False`, and treat them both the same way after init time? If so, I don't really have any objections to that if that's more in line with other languages. Just want to make sure we're thinking about this the same way first

Reply:
Sorry for the confusion. I went through the diagram again and the current implementation is correct.
Resolving the conversation.